sys/drv/bcm2835_pic.c

```diff
@@ -0,0 +1,204 @@
+#define KL_LOG KL_DEV
+#include <sys/bus.h>
+#include <sys/devclass.h>
+#include <sys/errno.h>
+#include <sys/fdt.h>
+#include <sys/interrupt.h>
+#include <sys/klog.h>
+#include <sys/libkern.h>
+#include <dev/bcm2835reg.h>
+
+/*
+ * located at BCM2835_ARMICU_BASE
+ * accessed by BCM2835_PERIPHERALS_BASE NOT by BCM2835_PERIPHERALS_BASE_BUS
+ * 32 GPU0 interrupts -- 0x204 offset from address; 0 in table
+ * 32 GPU1 interrupts -- 0x208 offset from address; 32 in table
+ * 32 basic interrupts -- 0x200 offset from address; 64 in table
+ */
+
+/* NOTE: the values of the following enum must not be modified! */
+typedef enum bcm2835_intrtype {
+  BCM2835_INTRTYPE_BASIC,
+  BCM2835_INTRTYPE_GPU0,
+  BCM2835_INTRTYPE_GPU1,
+  BCM2835_INTRTYPE_CNT
+} bcm2835_intrtype_t;
+
+#define BCM2835_NIRQPERTYPE (BCM2835_NIRQ / BCM2835_INTRTYPE_CNT)
+
+typedef struct bcm2835_pic_state {
+  rman_t rm;
+  resource_t *mem;
+  resource_t *irq;
+  intr_event_t *intr_event[BCM2835_NIRQ];
+} bcm2835_pic_state_t;
+
+#define in4(addr) bus_read_4(bcm2835_pic->mem, (addr))
+#define out4(addr, val) bus_write_4(bcm2835_pic->mem, (addr), (val))
+
+static void bcm2835_pic_disable_irq(intr_event_t *ie) {
+  bcm2835_pic_state_t *bcm2835_pic = ie->ie_source;
+  unsigned irq = ie->ie_irq;
+
+  bus_size_t reg = BCM2835_INTC_IRQBDISABLE;
+  if (irq < BCM2835_INT_GPU1BASE)
+    reg = BCM2835_INTC_IRQ1DISABLE;
+  else if (irq < BCM2835_INT_BASICBASE)
+    reg = BCM2835_INTC_IRQ2DISABLE;
+
+  uint32_t en = in4(reg);
+  out4(reg, en & ~(1 << irq));
+}
+
+static void bcm2835_pic_enable_irq(intr_event_t *ie) {
+  bcm2835_pic_state_t *bcm2835_pic = ie->ie_source;
+  unsigned irq = ie->ie_irq;
+
+  bus_size_t reg = BCM2835_INTC_IRQBENABLE;
+  if (irq < BCM2835_INT_GPU1BASE)
+    reg = BCM2835_INTC_IRQ1ENABLE;
+  else if (irq < BCM2835_INT_BASICBASE)
+    reg = BCM2835_INTC_IRQ2ENABLE;
+
+  uint32_t en = in4(reg);
+  out4(reg, en | (1 << irq));
```

Comment:
Not for this PR but for future:
It's time to do bitops in right way - once for whole codebase. 
In lib/libc/gen/fts.c we have following macros: `CLR`, `ISSET`, `SET`. I would opt for creating functions / macros in mimiker.h for that purpose and use them everywhere in kernel code.

Reply:
Great idea. We have a number of use cases throughout device drivers and each of them implements the operations on its own (in exactly the same way).

Reply:
I agree. Since device register access is a common use case we should provide easy to use set of macros to access them. But let's merge this PR quickly since already #1277 depends on it.